### PR TITLE
Make action container log reading more robust

### PR DESCRIPTION
**Pull request by sven-lange-last**
This PR fixes a small defect in action container log file reading in the invoker and addresses a premature file end when reading the log file.

### CODE REVIEW

File: core/invoker/src/main/scala/whisk/core/container/ContainerUtils.scala [hunk 1]
```diff
@@ -154,7 +154,7 @@ trait ContainerUtils {
      * Reads the contents of the file at the given position.
      * It is assumed that the contents does exist and that region is not changing concurrently.
      */
-    def getDockerLogContent(containerHash: ContainerHash, start: Long, end: Long, mounted: Boolean)(implicit transid: TransactionId): Array[Byte] = {
+    def getDockerLogContent(containerHash: ContainerHash, start: Long, end: Long, mounted: Boolean)(implicit transid: TransactionId): java.nio.ByteBuffer = {
```
Comment (markusthoemmes) [inline core/invoker/src/main/scala/whisk/core/container/ContainerUtils.scala]: `import java.nio.ByteBuffer` and remove the qualifications please.

File: core/invoker/src/main/scala/whisk/core/container/ContainerUtils.scala [hunk 2]
```diff
@@ -164,14 +164,19 @@ trait ContainerUtils {
...
             while (remain > 0) {
                 val read = channel.read(buffer)
-                if (read > 0)
-                    remain = read - read.toInt
+                if (read >= 0) {
+                    remain -= read
+                } else {
+                    // read < 0 means premature end of file => terminate loop
+                    remain = 0
+                }
```
Comment (markusthoemmes) [inline core/invoker/src/main/scala/whisk/core/container/ContainerUtils.scala]: Per personal discussion: We'll need a test that checks for this condition (and shows "wrong" behavior before the change)

File: core/invoker/src/main/scala/whisk/core/container/ContainerUtils.scala (hunk 1 above)
Comment (sven-lange-last) [inline core/invoker/src/main/scala/whisk/core/container/ContainerUtils.scala]: Changed as advised.

File: core/invoker/src/main/scala/whisk/core/container/ContainerUtils.scala (hunk 2 above)
Comment (rabbah) [inline core/invoker/src/main/scala/whisk/core/container/ContainerUtils.scala]: Hmm. I'm not seeing my comments here. I had asked for the same. I also explained there is a retry at the outer level; let me check where my comment went before rewriting.

Comment (rabbah) [inline core/invoker/src/main/scala/whisk/core/container/ContainerUtils.scala]: I don't know what happened to my comment. So here goes again: This log reader runs before tearing down a container to save the logs from an action container to a file for post inspection. Should we just nuke this code? There are also no tests for this code in general. This is not the log reader we use for activation log extraction.

Comment (rabbah) [inline core/invoker/src/main/scala/whisk/core/container/ContainerUtils.scala]: I posted and deleted two comments this morning as I try to remember what my concern was previously when I reviewed this PR. I can neither find that comment or remember exactly what it was but there are two high level bits. One, same as Markus' - test for the corner case? Second, looking back at the code, `read()` should not return -1 since we are reading a pre-determined number of bytes `pool.getDockerLogContent(con.containerId, con.lastLogSize, size, runningInContainer)`.

Comment (sven-lange-last) [inline core/invoker/src/main/scala/whisk/core/container/ContainerUtils.scala]: We are only talking about edge cases here.

Current line 168 is wrong and would set `remain`to zero whenever `channel.read()` was able to read something - in other words: almost always. This bug actually delegates all problems with reading to the caller of `getDockerLogContent()` eating up the retry count.

Two different cases are possible:

1. It _could_ happen that `channel.read()` is not able to fill the whole buffer in the first attempt, i.e. `0 < read < remain`. The bug in line 168 causes the loop to terminate. The calling function cannot find the sentinel marker, will drop the read buffer and will retry - invoking `getDockerLogContent()` again. This condition may arise if the log file is large and a lot of I/O is going on.

2. It _could_ happen that Docker truncates the log files after the its size has been determined by the caller of `getDockerLogContent()`. In case of full truncation, the loop is repeated until `channel.read()` returns a positive value - potentially blocking forever.

Both cases are expected to rarely happen in the wild - but if they happen, they can degrade performance considerably. And the fix is easy.

My recent commit `1e1494b847f4e0548d66db9d2108b3b514b5e082` adds unit tests.

Comment (rabbah) [inline core/invoker/src/main/scala/whisk/core/container/ContainerUtils.scala]: read - read 😳Thanks for pointing that out.

File: tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        fixture.writer.write(content)
+        fixture.writer.flush()
+        fixture.file.length
+    }
+
+    it should "tolerate an empty log file" in { fixture =>
+        val logText = ""
+        val size = writeLogFile(fixture, logText)
+
+        val buffer = fixture.cu.getDockerLogContent(containerHash = ContainerHash.fromString("0123"), start = 0, end = size, mounted = false)
```
Comment (markusthoemmes) [inline tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala]: externalize the default ContainerHash. We don't care about it anyways.

File: tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        val logContent = new String(buffer.array, buffer.arrayOffset, buffer.position, StandardCharsets.UTF_8)
+
+        logContent shouldBe logText
+        stream.size() shouldBe 0
```
Comment (markusthoemmes) [inline tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala]: `stream should have size 0`

File: tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+    it should "read a log file portion" in { fixture =>
+        val count = 4
+        val logText = (1 to count).map(i => s"Line ${i}\n").mkString("")
+        val size = writeLogFile(fixture, logText)
+        val quarter = count / 4
+        val qlen = logText.length / 4 // length of a quarter
+        val from = qlen // start reading from second quarter
+        val to = from + 2 * qlen // read second and third quarter
+        val expectedText = logText.substring(from, to)
```
Comment (markusthoemmes) [inline tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala]: Why so complex? Why not just:

```
val logText = "hello scala, this is quite a nice thing we got running here, eh?"
val from  = 4
val to = 10

...

logContent shouldBe logText.substring(from, to)
```

File: tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+        val fixture = FixtureParam(file, writer, cu)
+
+        try {
+            withFixture(test.toNoArgTest(fixture))
+        } finally {
+            writer.close()
+            file.delete()
+        }
+    }
```
Comment (markusthoemmes) [inline tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala]: This looks a bit overcomplicated + magic happens (methodnames need to match). Why not take a simpler approach like:

```
def createFile(content: String): File = //create your file and write the string to it
```

and then for each test you just initialize the class under test with the respective file:

```
it should .... in {
    val file = createFile("text")
    val utils = new TestContainerUtils(file)
}
```

I'd say that's more readable and involves far less magic. thoughts?

File: tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        fixture.cu.logFile = new File("/nonsense")
+
+        val buffer = fixture.cu.getDockerLogContent(containerHash = ContainerHash.fromString("0123"), start = 0, end = 1, mounted = false)
+
+        buffer.capacity() shouldBe 0
+        logLines.head should include("getDockerLogContent failed")
+    }
+}
+
+class ContainerUtilsTester(var logFile: File)(implicit val logging: Logging) extends ContainerUtils {
```
Comment (markusthoemmes) [inline tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala]: If you apply the changes suggested above, this can become a `val` or even easier: a private member.

File: tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+        val buffer = fixture.cu.getDockerLogContent(containerHash = ContainerHash.fromString("0123"), start = from, end = to, mounted = false)
+        val logContent = new String(buffer.array, buffer.arrayOffset, buffer.position, StandardCharsets.UTF_8)
+
+        logContent shouldBe expectedText
+        stream.size() shouldBe 0
+    }
+
+    it should "tolerate premature end of log file" in { fixture =>
+        val logText = (1 to 2).map(i => s"Line ${i}\n").mkString("")
```
Comment (markusthoemmes) [inline tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala]: If you want to write lines, why not do

```
(1 to 2).map(i => s"Line ${i}").mkString("\n")
```

File: tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+/**
+ * Unit tests for ContainerPool and, by association, Container and WhiskContainer.
+ */
+@RunWith(classOf[JUnitRunner])
+class ContainerUtilsTests extends FlatSpec
+    with BeforeAndAfter
+    with Matchers
+    with StreamLogging
+    with TimeLimitedTests {
```
Comment (markusthoemmes) [inline tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala]: Why is this needed?

File: tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+
+    val timeLimit = 100 millis
```
Comment (markusthoemmes) [inline tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala]: Unused value.

File: tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala (hunk 9 above)
Comment (sven-lange-last) [inline tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala]: Time limitation of tests is not strictly needed at the moment. The tests perform real I/O instead of mocked I/O so that a time limitation may prevent too long runtimes when running this test suite as part of a full build.

In addition, my initial assumption was to prepare unit tests for the original code. With the bug fixed by this PR, at least one unit test would have looped such that a time limitation had been necessary to prevent excessive test runtime.

File: tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala (hunk 10 above)
Comment (sven-lange-last) [inline tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala]: See http://doc.scalatest.org/1.8/org/scalatest/concurrent/TimeLimitedTests.html.

`timeLimit` overrides a `def` from trait `org.scalatest.concurrent.TimeLimitedTests` activating the specified time limit for each single test.

Comment (sven-lange-last) [inline tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala]: I added a comment to this line to point this out.

File: tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala (hunk 9 above)
Comment (sven-lange-last) [inline tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala]: Just discussed with @markusthoemmes - removing the time limit because it isn't required at the moment.

File: tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala (hunk 6 above)
Comment (sven-lange-last) [inline tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala]: I was using fixtures because they provide a central place for setting and - particularly - cleaning up. The fixture makes sure that temporary files are closed and removed after each test.

All this is also possible with the approach you suggest above - but requires a `try ... finally`bracket around each single test.

File: tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala (hunk 3 above)
Comment (sven-lange-last) [inline tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala]: Makes sense.

File: tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala (hunk 4 above)
Comment (sven-lange-last) [inline tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala]: Good proposal - thanks. Changed.

File: tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala (hunk 8 above)
Comment (sven-lange-last) [inline tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala]: Your proposal does only add `\n` between the strings included in the range. In other words: with your proposal, the last line won't be followed by a `\n`.

File: tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala (hunk 7 above)
Comment (sven-lange-last) [inline tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala]: I understand... When keeping the fixture approach, test `provide an empty result on failure` will be invoked with a fixture that isn't really used...

At the same time, said test overrides the `logFile` to a file that doesn't exist to provoke a failure.

File: tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala (hunk 5 above)
Comment (sven-lange-last) [inline tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala]: Done

File: tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala (hunk 8 above)
Comment (rabbah) [inline tests/src/test/scala/whisk/core/container/test/ContainerUtilsTests.scala]: There's a mkstring variant with separators at start, middle and end.

**Comment by sven-lange-last** [comment]
PG2 - 1488 running

**Comment by rabbah** [comment]
@markusthoemmes rebased this to master; pg4/552 running since I couldn't verify the above noted pg.

**Comment by rabbah** [comment]
pg 4/552 🔵

**Comment by andrewnet** [comment]
Hi, can anyone merge this fix into master? I am still facing the problem and would like to give it a try once this is pulled in. I would do it if I have write access.

**Comment by rabbah** [comment]
We've actually switched to a new implementation of the invoker which doesn't have this bug. Are you seeing the issue still in bluemix?

**Comment by rabbah** [comment]
This is the code that's now active https://github.com/apache/incubator-openwhisk/blob/master/core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala. The code affected by this pr is effectively deadcode we plan on removing.

**Comment by andrewnet** [comment]
Yes, I am still seeing logs got truncated on YP-QA environment.

**Comment by andrewnet** [comment]
I have opened a defect https://github.com/apache/incubator-openwhisk/issues/2548 to report the issue separately since this defect is out dated as @rabbah.

**Comment by markusthoemmes** [comment]
@sven-lange-last want to port this to the new folderstructure?

**Comment by rabbah** [comment]
@sven-lange-last @markusthoemmes is this still relevant or now out of date and should be closed?

**Comment by markusthoemmes** [comment]
I think we can safely close this, the code is gone anyway.
